### PR TITLE
Add async client support

### PR DESCRIPTION
**Pull request by balamurugana**
This commit comes with
* Adds `MinioAsyncClient` class contains APIs returning `CompletableFuture` to execute them in future.
* All APIs in `MinioClient` underneath uses APIs in `MinioAsyncClient` as base.
* All low level S3 APIs in `S3Base` class return `CompletableFuture` and older APIs are deprecated.

Signed-off-by: Bala.FA <bala@minio.io>

### CODE REVIEW

**Comment by harshavardhana** [comment]
is this a breaking change? @balamurugana

**Comment by balamurugana** [comment]
> is this a breaking change? @balamurugana 

No. The existing functionalities work as is in addition to async APIs.

File: api/src/main/java/io/minio/ListObjectsV1Response.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
 import io.minio.messages.ListBucketResultV1;
 import okhttp3.Headers;
 
-/** Response class of {@link MinioClient#listObjectsV1}. */
+/** Response class of {@link S3Base#listObjectsV1}. */
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/ListObjectsV1Response.java:22]:
```suggestion
/** Response class of {@link S3Base#listObjectsV1Async}. */
```

File: api/src/main/java/io/minio/ListObjectsV2Response.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 import io.minio.messages.ListBucketResultV2;
 import okhttp3.Headers;
 
-/** Response class of {@link MinioClient#listObjectsV2}. */
+/** Response class of {@link S3Base#listObjectsV2}. */
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/ListObjectsV2Response.java:22]:
```suggestion
/** Response class of {@link S3Base#listObjectsV2Async}. */
```

File: api/src/main/java/io/minio/ListenBucketNotificationArgs.java [hunk 3]
```diff
@@ -19,7 +19,10 @@
 import java.util.Arrays;
 import java.util.Objects;
 
-/** Argument class of {@link MinioClient#listenBucketNotification}. */
+/**
+ * Argument class of {@link MinioClient#listenBucketNotification} and {@link
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/ListenBucketNotificationArgs.java:23]:
```suggestion
 * Argument class of {@link MinioAsyncClient#listenBucketNotification} and {@link
```

File: api/src/main/java/io/minio/StatObjectResponse.java [hunk 4]
```diff
@@ -26,7 +26,7 @@
 import java.util.Map;
 import okhttp3.Headers;
 
-/** Response of {@link MinioClient#statObject}. */
+/** Response of {@link MinioAsyncClient#statObject} and {@link MinioClient#statObject}. */
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/StatObjectResponse.java:29]:
```suggestion
/** Response of {@link S3Base#statObjectAsync} */
```

**Review by anjalshireesh — CHANGES_REQUESTED** [review]

File: api/src/main/java/io/minio/StatObjectResponse.java (hunk 4 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/StatObjectResponse.java]: done

File: api/src/main/java/io/minio/ListenBucketNotificationArgs.java (hunk 3 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/ListenBucketNotificationArgs.java]: done

File: api/src/main/java/io/minio/ListObjectsV2Response.java (hunk 2 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/ListObjectsV2Response.java]: done

File: api/src/main/java/io/minio/ListObjectsV1Response.java (hunk 1 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/ListObjectsV1Response.java]: done

**Review by anjalshireesh — APPROVED** [review]
LGTM, one minor suggestion

File: api/src/main/java/io/minio/S3Base.java [hunk 5]
```diff
@@ -200,6 +205,66 @@ protected Multimap<String, String> newMultimap(Multimap<String, String> map) {
...
+
+    if (ex instanceof ServerException) {
+      throw (ServerException) ex;
+    }
+
+    if (ex instanceof XmlParserException) {
+      throw (XmlParserException) ex;
+    }
+
+    throw new RuntimeException(ex);
```
[the remark this answers is not in the document]
Comment (harshavardhana) [reply api/src/main/java/io/minio/S3Base.java:265]: PTAL @balamurugana

**Review by harshavardhana — APPROVED** [review]

**Review by anjalshireesh — APPROVED** [review]

**Pull request merged**